core/package-graph/index.js

```diff
@@ -238,19 +350,71 @@ class PackageGraph extends Map {
       currentNode.localDependents.forEach(visits([currentName]));
     });
 
-    if (cyclePaths.size) {
-      const cycleMessage = ["Dependency cycles detected, you should fix these!"]
-        .concat(Array.from(cyclePaths).map(cycle => cycle.join(" -> ")))
-        .join("\n");
+    reportCycles(Array.from(cyclePaths, cycle => cycle.join(" -> ")), rejectCycles);
+
+    return [cyclePaths, cycleNodes];
+  }
+
+  /**
+   * Returns the cycles of this graph. If two cycles share some elements, they will
+   * be returned as a single cycle.
+   *
+   * @param {!boolean} rejectCycles Whether or not to reject cycles
+   * @returns Set<PackageGraphCollapsedNode>
+   */
+  collapseCycles(rejectCycles) {
+    const cyclePaths = [];
+    const nodeToCycle = new Map();
+    const cycles = new Set();
+
+    const walkStack = [];
 
-      if (rejectCycles) {
-        throw new ValidationError("ECYCLE", cycleMessage);
+    const visits = (baseNode, dependentNode) => {
+      if (nodeToCycle.has(baseNode)) {
+        return;
       }
 
-      log.warn("ECYCLE", cycleMessage);
-    }
+      let topLevelDependent = dependentNode;
+      while (nodeToCycle.has(topLevelDependent)) {
+        topLevelDependent = nodeToCycle.get(topLevelDependent);
+      }
 
-    return [cyclePaths, cycleNodes];
+      if (
+        topLevelDependent === baseNode ||
+        (topLevelDependent.isCycle && topLevelDependent.has(baseNode.name))
+      ) {
+        const cycle = new PackageGraphCollapsedNode();
+
+        walkStack.forEach(nodeInCycle => {
+          nodeToCycle.set(nodeInCycle, cycle);
+          cycle.addPackage(nodeInCycle);
+          cycles.delete(nodeInCycle);
+        });
+
+        cycles.add(cycle);
+        cyclePaths.push(cycle.toString());
+
+        return;
+      }
+
+      if (walkStack.indexOf(topLevelDependent) === -1) {
+        // eslint-disable-next-line no-use-before-define
+        visitWithStack(baseNode, topLevelDependent);
+      }
+    };
+
+    const visitWithStack = (baseNode, currentNode = baseNode) => {
+      walkStack.push(currentNode);
+      currentNode.localDependents.forEach(visits.bind(null, baseNode));
```

Comment:
The old implementation used an arrow function (`dependentNode => visits(baseNode, dependentNode)`), but `.bind` makes it way easier to understand the stack while debugging since the intermediate function is elided.